src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -2624,20 +2624,27 @@ void MacroAssembler::cmpxchgptr(Register oldv, Register newv, Register addr, Reg
   // oldv holds comparison value
   // newv holds value to write in exchange
   // addr identifies memory word to compare against/update
-  Label retry_load, nope;
-  bind(retry_load);
-  // Load reserved from the memory location
-  load_reserved(tmp, addr, int64, Assembler::aqrl);
-  // Fail and exit if it is not what we expect
-  bne(tmp, oldv, nope);
-  // If the store conditional succeeds, tmp will be zero
-  store_conditional(tmp, newv, addr, int64, Assembler::rl);
-  beqz(tmp, succeed);
-  // Retry only when the store conditional failed
-  j(retry_load);
-
-  bind(nope);
-  membar(AnyAny);
+  if (UseZacas) {
+    mv(tmp, oldv);
+    atomic_cas(tmp, newv, addr, Assembler::int64, Assembler::aq, Assembler::rl);
+    beq(tmp, oldv, succeed);
```

Comment:
That dmb is not present in the AArch64 port because we want a release when the CAS fails, because if it fails nothing was stored, so there is literally nothing for a subsequent load from that address to synchronize with.

It's there because of this re-ordering:
```


	<Access [A]>

	// atomic_op (B)
1:	ldaxr	x0, [B]		// Exclusive load with acquire
	<op(B)>
	stlxr	w1, x0, [B]	// Exclusive store with release
	cbnz	w1, 1b

	<Access [C]>

It doesn't forbid orderings such as

Load [B] -> Load [C] -> Store [A] -> Store [B]
```

[See here](https://mail.openjdk.org/pipermail/aarch64-port-dev/2014-February/000706.html)

The Arm memory model has been strengthened, and this reasoning looks a bit shaky today. At the time we did not know if any of the usages of `cmpxchgptr`required "full barrier" semantics, so we put a full barrier in for safety's sake.

Reply:
Wow, thanks for finding that history. It's very helpful for us to understand the existence of this barrier.

Reply:
I have seen code, do not find it anymore in code base, which use cmpxchg for Atomic::release_store_fence().
(which is "new");
```
Atomic::store(_x, true);
Atomic::store_release_fence(_one_way_barrier, true);
bool z = Atomic::load(_z);
```
But it is written as:
```
Atomic::store(_x, true);
Atomic::cmpxchg(_one_way_barrier, false, true);
bool z = Atomic::load(_z);
```
Where load Z happens after store X. But without release the store may happens after.

As we say:
"All of the atomic operations that imply a read-modify-write action guarantee a two-way memory barrier across that operation."

I have look over the cmpxchg uses and not releasing 'seems' okay at first glance.
But I think before this line changes we should release on failed CAS.

Reply:
Note I mention the hotspot runtime atomic here because cmpxchgptr is used to manipulate the markword which follows hotspot atomics, not Java volatiles.

Reply:
> I have seen code, do not find it anymore in code base, which use cmpxchg for Atomic::release_store_fence(). (which is "new");
> 
> ```
> Atomic::store(_x, true);
> Atomic::store_release_fence(_one_way_barrier, true);
> bool z = Atomic::load(_z);
> ```
> 
> But it is written as:
> 
> ```
> Atomic::store(_x, true);
> Atomic::cmpxchg(_one_way_barrier, false, true);
> bool z = Atomic::load(_z);
> ```
> 
> Where load Z happens after store X. But without release the store may happens after.

Yuck. I hope that is never going to return.

I think all such usages of cmpxchg were replaced by an atomic add of zero into the stack, which works on x86. 

> As we say: "All of the atomic operations that imply a read-modify-write action guarantee a two-way memory barrier across that operation."

Well, that's for the HotSpot internal atomics, rather than for Java code. And given that this particular version of `cmpxchg` is only used in a few places it suffices to see what they need. 

> I have look over the cmpxchg uses and not releasing 'seems' okay at first glance. But I think before this line changes we should release on failed CAS.

I have never agreed with that thinking. We should find the callers of CAS that depend on such a side effect and fix them. The problem is that no-one wants to do it. There may no longer be any such cases.